src/Rules/Library.hs

```diff
@@ -1,116 +1,96 @@
-import Hadrian.Haskell.Cabal.Parse (parseCabalPkgId)
+import Hadrian.Package.Type
 
 import Base
 import Context
 import Expression hiding (way, package)
 import Flavour
 import GHC.Packages
 import Oracles.ModuleFiles
+import Oracles.Setting (libsuf)
 import Rules.Gmp
 import Settings
 import Target
 import Utilities
 
+import Data.Functor
 import qualified System.Directory as IO
+import qualified Text.Parsec      as Parsec
 
-archive :: Way -> String -> String
-archive way pkgId = "libHS" ++ pkgId ++ (waySuffix way <.> "a")
-
--- TODO: This comment is rather vague, make it more precise by listing what
--- exactly gets built and moved where, referencing the corresponding rules.
--- | Building a library consist of building the artefacts, copying it somewhere
--- with Cabal, and finally registering it with the compiler via Cabal in the
--- package database. We assume rules to build all the package artefacts, and
--- provide rules for the library artefacts.
-library :: Context -> Rules ()
-library context@Context{..} = do
-    root <- buildRootRules
-    pkgId <- case pkgCabalFile package of
-        Just file -> liftIO $ parseCabalPkgId file
-        Nothing   -> return $ pkgName package
-
-    root -/- libDir context -/- pkgId -/- archive way pkgId %> \_ ->
-        need =<< mapM (\pkgId -> packageDbPath stage <&> (-/- pkgId <.> "conf")) [pkgId]
+-- * Library 'Rules'
 
-libraryObjects :: Context -> Action [FilePath]
-libraryObjects context@Context{..} = do
-    hsObjs    <- hsObjects    context
-    nonHsObjs <- nonHsObjects context
+libraryRules :: Rules ()
+libraryRules = do
+  root <- buildRootRules
 
-    -- This will create split objects if required (we don't track them
-    -- explicitly as this would needlessly bloat the Shake database).
-    need $ nonHsObjs ++ hsObjs
+  root -/- "//libHS*-*.dylib" %> buildDynamicLibUnix root "dylib"
+  root -/- "//libHS*-*.so"    %> buildDynamicLibUnix root "so"
+  root -/- "//*.a" %> buildStaticLib root
+  priority 2 $ root -/- "//HS*-*.o" %> buildGhciLibO root
 
-    split <- interpretInContext context =<< splitObjects <$> flavour
-    let getSplitObjs = concatForM hsObjs $ \obj -> do
-            let dir = dropExtension obj ++ "_" ++ osuf way ++ "_split"
-            contents <- liftIO $ IO.getDirectoryContents dir
-            return . map (dir -/-) $ filter (not . all (== '.')) contents
+-- * 'Action's for building libraries
+
+-- | Build a static library ('LibA') under the given build root, whose
+--   path is the second argument.
+buildStaticLib :: FilePath -> FilePath -> Action ()
+buildStaticLib root archivePath = do
+  l@(BuildPath _ stage _ (LibA pkgname _ way))
+    <- parsePath (parseBuildLibA root)
+                 "<.a library (build) path parser>"
+                 archivePath
+  let context = libAContext l
+  objs <- libraryObjects context
+  asuf <- libsuf way
+  let isLib0 = ("//*-0" ++ asuf) ?== archivePath
+  removeFile archivePath
+  if isLib0
```

Comment:
This is dead code, which was dropped in: 

https://github.com/snowleopard/hadrian/commit/05fbe8b36da146aae24c53455131b5b5618ee849

Please migrate the changes here.

Reply:
Done in upcoming commit.